### DISCUSSION TITLE
How to setup ethernet driver to use UDP/IPv4 protocol

### DISCUSSION THREAD

**Discussion by HuyBui1327** [zephyrproject-rtos/zephyr#60985] [Q&A]
**My purpose**: 
- I'm trying to create TFTP client application using this lib `zephyr\subsys\net\lib\tftp\tftp_client.c` since I cannot find any TFTP sample app in **Zephyr v3.1.0**. 
- Because TFTP uses UDP/IPv4 protocol to transfer data. So, firstly I created the app to send UDP/IPv4 packet to server. 

**My setup**:
- My board directly connects to my laptop through Ethernet cable. 
- I use WireShark to capture the packet.

**My result**:
- Problem: The captured packets from my board (2e:09:0a:06:f1:67) to my laptop (Lux_share_10:24:a5) are using ARP protocol. 
![image](https://github.com/zephyrproject-rtos/zephyr/assets/131848739/c7dcc781-0338-4310-b4cb-13706035aea6)

**Debug**: In `zephyr\subsys\net\l2\ethernet\ethernet.c`
- To change ethernet header prototype manually, I have changed line 700 to 705:
```
/* Then set the ethernet header.
*/
if (!ethernet_fill_header(ctx, pkt, ptype)) {
	ret = -ENOMEM;
	goto error;
}
```
**to** 
```
/* Then set the ethernet header.
*/
ptype = htons(NET_ETH_PTYPE_IP);
if (!ethernet_fill_header(ctx, pkt, ptype)) {
	ret = -ENOMEM;
	goto error;
}
```
but the captured packets still have ethernet header prototype 0x806(ARP). Does func `ethernet_fill_header()` actually set ethernet header?

**My question**:
- Do I miss any step in my source code to enable UDP/IPv4 protocol?
- Do I enable the right CONFIG option in prj.conf?
Thank you in advance.

**My source code**:
- main.c

```
#include <zephyr/zephyr.h>
#include <ztest.h>

extern void test_ethernet(void);

void test_main(void)
{
	ztest_test_suite(eth_test,
			ztest_unit_test(test_ethernet));
	ztest_run_test_suite(eth_test);
}
```
- test_ethernet.c

```
#include <zephyr/logging/log.h>
LOG_MODULE_REGISTER(eth_sample, LOG_LEVEL_DBG);

#include <zephyr/zephyr.h>
#include <errno.h>
#include <stdio.h>

// #include <zephyr/posix/sys/socket.h>
#include <zephyr/net/ethernet.h>
#include <zephyr/net/tftp.h>

#define STACK_SIZE 1024

#define TFTP_SERVER_PORT 2023
#define INADDR_SERVER  "169.254.135.88"
#define INADDR_CLIENT  "169.254.135.86"


void test_ethernet(void)
{
	char client_message[11]="Hi, Server";
	char buf[15] = "00.00.000.000";

	int sock = -1;
	struct sockaddr_in dst = { 0 };
	struct in_addr inaddr_src = { 0 };

	dst.sin_family = AF_INET;
	dst.sin_port = htons(TFTP_SERVER_PORT);
	inet_pton(dst.sin_family, INADDR_SERVER, &dst.sin_addr);
	inet_ntop(dst.sin_family, &dst.sin_addr, buf, sizeof(buf));
	printf("SERVER inet_ntop: %s\n", buf);


	inet_pton(dst.sin_family, INADDR_CLIENT, &inaddr_src);
	inet_ntop(dst.sin_family, &inaddr_src, buf, sizeof(buf));
	printf("CLIENT inet_ntop: %s\n", buf);

	/* Get network interface */
        struct net_if* ptrNetIf = net_if_get_first_by_type(&NET_L2_GET_NAME(ETHERNET));

	/* Add a IPv4 address to an interface */
	net_if_ipv4_addr_add(ptrNetIf, &inaddr_src, NET_ADDR_MANUAL, 0);

	sock = socket(dst.sin_family, SOCK_DGRAM, IPPROTO_UDP);
	if (sock < 0) {
		LOG_ERR("Failed to create UDP socket (IPv4): %d", errno);
		return -errno;
	}

	/* Send the message to server */
        if(sendto(sock, client_message, sizeof(client_message), 0,
         (struct sockaddr*)&dst, sizeof(struct sockaddr_in)) < 0) {
                 LOG_ERR("Unable to send message\n");
                 return;
        }
	else {
		LOG_INF("Have sended message\n");
	}

}
```

**My prj.conf files**:
 ```
CONFIG_ZTEST=y

# Generic networking options
CONFIG_ETH_MY_BOARD=y
CONFIG_NET_L2_ETHERNET=y
CONFIG_NETWORKING=y
CONFIG_NET_MGMT=y
CONFIG_NET_L2_ETHERNET_MGMT=y
CONFIG_TEST_RANDOM_GENERATOR=y
CONFIG_NET_UDP=y
CONFIG_NET_TCP=n
CONFIG_NET_IPV6=n
CONFIG_NET_IPV4=y
CONFIG_NET_MAX_CONTEXTS=10
CONFIG_PTP_CLOCK=n

# Packet socket configuration
CONFIG_NET_SOCKETS=y
CONFIG_NET_SOCKETS_PACKET=y

# Destination PC MAC address
CONFIG_NET_SAMPLE_DESTINATION_ADDR="60:7d:09:10:24:a5"

# Board MAC address
CONFIG_ETH_MY_BOARD_MAC_ADDR="2e:09:0a:06:f1:67"

# Configuration for MY_BOARD Ethernet driver
CONFIG_ETH_MY_BOARD_RX_BUF_TOTAL=20
CONFIG_ETH_MY_BOARD_TX_BUF_TOTAL=10
CONFIG_ETH_MY_BOARD_RX_BUF_SIZE=256
CONFIG_ETH_MY_BOARD_TX_BUF_SIZE=256
CONFIG_ETH_MY_BOARD_SPEED=100
CONFIG_PHY_INIT_PRIORITY=80
CONFIG_ETH_INIT_PRIORITY=70

# Log level
CONFIG_ETHERNET_LOG_LEVEL_INF=y
CONFIG_PHY_LOG_LEVEL_INF=y

# Logging
CONFIG_LOG=y
CONFIG_NET_LOG=y
CONFIG_NET_STATISTICS=y
CONFIG_PRINTK=y

CONFIG_NET_SOCKETS_LOG_LEVEL_DBG=n
CONFIG_NET_CONTEXT_LOG_LEVEL_DBG=y
CONFIG_NET_CORE_LOG_LEVEL_DBG=n
CONFIG_NET_CONN_LOG_LEVEL_DBG=n
CONFIG_NET_IF_LOG_LEVEL_DBG=y
CONFIG_NET_L2_ETHERNET_LOG_LEVEL_DBG=y
CONFIG_ETHERNET_LOG_LEVEL_DBG=y
CONFIG_NET_PKT_LOG_LEVEL_ERR=n

CONFIG_TFTP_LIB=y
CONFIG_POSIX_API=n
CONFIG_TFTP_LOG_LEVEL_DBG=y

CONFIG_NET_SHELL=y
```

**Comment by RomainPelletant** [comment]
Hi Huybui1327,

You may need IP address attribution (DHCP) or static one to make L3/L4 layer enabled.

To investigate on network, it could be helpful to see output of netshell : net iface and see IP related data

**Reply by HuyBui1327** [reply]
Hi @RomainPelletant,

Thanks for your response and sorry to bother you.

> You may need IP address attribution (DHCP) or static one to make L3/L4 layer enabled.

In file `test_ethernet.c`, I have set static IP address to make L3/L4 layer enabled. Do I miss any steps? 
The Server and Client have the same network class so do I need to enable DHCP? If yes, I should refer file `samples/net/dhcpv4_client/main.c` right?
```
#define INADDR_SERVER  "169.254.135.88"
#define INADDR_CLIENT  "169.254.135.86"
...
       /* Get network interface */
        struct net_if* ptrNetIf = net_if_get_first_by_type(&NET_L2_GET_NAME(ETHERNET));

	/* Add a IPv4 address to an interface */
	net_if_ipv4_addr_add(ptrNetIf, &inaddr_src, NET_ADDR_MANUAL, 0);
...
```



> To investigate on network, it could be helpful to see output of netshell : net iface and see IP related data

Here is my net iface:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/131848739/743ed919-8484-41bf-9894-2fe3c0432a8a)

In my understanding, currently there is no TFTP sample app in Zephyr repo right?

**Reply by RomainPelletant** [reply]
You are right, there is no sample for TFTP at this time. And you don't need DHCP for static addressing.

IMO, before launching TFTP client, it could be useful to ensure network connection is ready to go before doing something on the network.
Take a look at net_mgmt.h : you can register to network events on L2, L3 and L4 layer.
To test it : add CONFIG_NET_MGMT_EVENT_MONITOR=y
Then you can see network event states using net shell.
Once you get "connected" (NET_EVENT_L4_CONNECTED) status, you can launch the tftp client. 

To use directly the TFTP Client library, you will need to instanciate a struct tftpc, implementing the callback and server address as sockaddr. Then call tftp_get or tftp_put as described in #include <zephyr/net/tftp.h>

This could be an opportunity to add a sample in Zephyr upstream :)

**Reply by HuyBui1327** [reply]
> This could be an opportunity to add a sample in Zephyr upstream :)

wish the luck for me :)
But I will give it a try and address you if any update.

Again, thank you so much for the advice.

**Reply by RomainPelletant** [reply]
@HuyBui1327 

Here is a [PR #61626](https://github.com/zephyrproject-rtos/zephyr/pull/61626) with code sample for TFTP client.
Not tested yet but it could be a good starting point for TFTP part.
